docs/upgrade-to-v3.md

```diff
@@ -33,6 +33,136 @@ The Microsoft Graph Java SDK now leverages azure-identity to improve the authent
+```
+
+##### Client credential flow
+
+Replace any of the following:
+
+```Java
+ClientCredentialProvider authProvider = new ClientCredentialProvider(CLIENT_ID, SCOPES, CLIENT_SECRET, TENANT_GUID, NATIONAL_CLOUD);
+IGraphServiceClient graphClient = GraphServiceClient.builder().authenticationProvider(authProvider).buildClient();
+```
+
+By
+
+``` Java
+final ClientSecretCredential clientSecretCredential = new ClientSecretCredentialBuilder()
+    .clientId(CLIENT_ID)
+    .clientSecret(CLIENT_SECRET)
+    .tenantId(TENANT_GUID)
+    .build();
+final TokenCredentialAuthProvider authProvider = new TokenCredentialAuthProvider(SCOPES, clientSecretCredential);
+final GraphServiceClient graphClient = GraphServiceClient.builder().authenticationProvider(authProvider).buildClient();
+```
+
+> Note: you can alternatively use **ClientCertificateCredential** to authenticate using a certificate instead of a secret. Using certificates for authentication was not previously supported.
+
+##### User name and password flow
+
+Replace any of the following:
+
+```Java
+UsernamePasswordProvider authProvider = new UsernamePasswordProvider(CLIENT_ID, SCOPES, USERNAME, PASSWORD, NATIONAL_CLOUD, TENANT, CLIENT_SECRET);
+IGraphServiceClient graphClient = GraphServiceClient.builder().authenticationProvider(authProvider).buildClient();
+```
+
+By
+
+``` Java
+final UsernamePasswordCredential usernamePasswordCredential = new UsernamePasswordCredentialBuilder()
+    .clientId(CLIENT_ID)
+    .username(USERNAME)
+    .password(PASSWORD)
+    .build();
+final TokenCredentialAuthProvider authProvider = new TokenCredentialAuthProvider(SCOPES, usernamePasswordCredential);
+final GraphServiceClient graphClient = GraphServiceClient.builder().authenticationProvider(authProvider).buildClient();
+```
+
+##### Device code flow
+
+> Note: the device code flow was previously not available, some customers might have implemented custom authentication providers in their application to bridge the feature gap. That custom code can now be replaced.
+
+Replace any of the following:
+
+```Java
+CustomBuiltDeviceCodeProvider authProvider = new CustomBuiltDeviceCodeProvider(CLIENT_ID); // code specific to the custom implementation
+IGraphServiceClient graphClient = GraphServiceClient.builder().authenticationProvider(authProvider).buildClient();
+```
+
+By
+
+``` Java
+final DeviceCodeCredential deviceCodeCredential = new DeviceCodeCredentialBuilder()
+    .clientId(CLIENT_ID)
+    .challengeConsumer(challenge -> {
+                        // lets user know of the challenge
+                        System.out.println(challenge.getMessage());
+                    })
+    .build();
+final TokenCredentialAuthProvider authProvider = new TokenCredentialAuthProvider(SCOPES, deviceCodeCredential);
+final GraphServiceClient graphClient = GraphServiceClient.builder().authenticationProvider(authProvider).buildClient();
+```
+
+##### Interwactive browser flow
```

Comment:
```suggestion
##### Interactive browser flow
```

Reply:
ahhh memory muscle blip from the first part of my life being on AZERTY keyboards....

Reply:
AZERTY! Reminds me of my internship days in France :)